### ISSUE TITLE
Client proceeds in handshake after receiving consecutive CertificateRequest messages

### ISSUE THREAD

**Issue by pfg666** [eclipse/tinydtls#78]
## Non-conformance Bug
**OS**  
Linux, Ubuntu 20
develop branch, commit c20c6323

**Expected behavior**
The client should not proceed with a handshake where it receives distinct consecutive CertificateRequest messages.

**Actual Behavior**
Our testing shows that the client upon receiving these messages proceeds with the handshake as normal; upon receiving ServerHelloDone, it generates the next flight of messages as can be seen in the capture below taken on my machine.

![etinydtls_multiple_certreq](https://user-images.githubusercontent.com/2325013/119117008-198c9080-ba29-11eb-8bd9-b057f4c84bc4.png)

Note that the later CertificateRequest is not a retransmission of the former. Its message_sequence value differs, so it should have been processed as a distinct handshake message.

**Steps to Reproduce**
I attached files necessary for reproduction using [DTLS-Fuzzer](https://github.com/assist-project/dtls-fuzzer/), a Java-based tool for testing DTLS libraries. Also included in the archive is a capture of the interaction show above. DTLS-Fuzzer requires  the JDK for Java 8. On Ubuntu, this can be installed  by running:
`sudo apt-get install openjdk-8-jdk`

Unpack the archive, `cd` to resulting folder and run `bash reproduce.sh`, while running an instance of Wireshark on the side. The reproduction script will: 
* setup eclipse TinyDTLS and DTLS-Fuzzer
* launch the TinyDTLS client using the 'tests/dtls-client utility;
* launch DTLS-Fuzzer to execute the input sequence 'test_sequence'  which exposes the bug.

Thanks!

[reproduction.tar.gz](https://github.com/eclipse/tinydtls/files/6521626/reproduction.tar.gz)

**Comment by mrdeep1** [comment]
@pfg666 With your script, I get the following failure (I had to do `sudo apt install maven` and then build everything from scratch)
````
13:38:18 [main] INFO : Main - Noticed that the first argument is a file. Processing it as an argument file.
13:38:18 [main] INFO : Main - Processing command state-fuzzer-client
13:38:18 [main] INFO : Main - Debug operation is engaged
13:38:18 [main] INFO : Main - Running test runner
13:38:23 [main] WARN : WorkflowConfigurationFactory - Unsupported key exchange algorithm: NULL, not creating ClientKeyExchange Message
13:38:24 [Thread-1] INFO : TlsSUL - Initializing transport handler (waiting ClientHello)
13:38:24 [main] ERROR: TlsSUL - Start 0
````

**Comment by pfg666** [comment]
Hi, the "ERROR"  line is a result of sloppy programming (DTLS-Fuzzer is a research prototype designed by sometimes careless developers).  It is not an error, but a badly posted message log. Thanks for reporting!

I will fix the logging as well as add mvn install in DTLS-Fuzzer's  'prepare.sh' script. Are you seeing any messages (DTLS or ICMP) as a result of running 'bash reproduce.sh' or is this the full log and it blocks (you have to do Ctrl+C)?  In case the latter is true, it means DTLS-Fuzzer is waiting for the client to connect (the script possibly started the TinyDTLS client too early). In 'reproduce.sh', change at line 98 the value supplied to `sleep` to something bigger then 3. 

Alternatively, in a separate terminal launch the Eclipse tinydtls client by running: ` etinydtls/tests/dtls-client -p 20000 localhost ` from the archive folder. Let me know if this works/if you get any interaction between the client and DTLS-Fuzzer.

**Comment by mrdeep1** [comment]
sleep 10 fixes the reproduction issue - thanks.  Now to think about handling correctly this situation.

**Comment by boaks** [comment]
For me the state-machine (`handle_handshake_msg`) seems to keep the mandatory states only, and so the optional states seems to be not checked for message duplicates.

e.g.:

client in `DTLS_STATE_WAIT_SERVERHELLODONE`, 
`DTLS_HT_SERVER_KEY_EXCHANGE` is optional for PSK, receiving that multiple times is not checked. Also keeping the optional states (e.g. `DTLS_STATE_WAIT_OPTIONAL_SERVERKEYEXCHANGE`) would solve it.
Same for `DTLS_HT_CERTIFICATE_REQUEST`, requires also a `DTLS_STATE_WAIT_OPTIONAL_CERTIFICATE_REQUEST`. That makes the state check a little more complicate, but prevents such errors reported here.

**Comment by mrdeep1** [comment]
@pfg666 PR #88 should fix this issue  - please confirm (thanks @boaks for the ServerKeyExchange hint - I could not find any other options handshake duplication issues).  Note that PR #87 will stop the following records being necessarily buffered.

**Comment by pfg666** [comment]
@mrdeep1 I can confirm that the problem is fixed by PR #88. The issue can be closed. BTW, I don't know what the etiquette is with respect to who should close the issue, whether it is the reporter or the devs. My leaning is to let the devs do the closing.

**Comment by mrdeep1** [comment]
Thanks.  It is worth waiting for PR #88 to be merged before closing this Issue.  I do not have the ability to close off any Issues (or do a merge) though.

**Comment by obgm** [comment]
Thanks @pfg666 @mrdeep1 and @boaks. I encourage the reporter to close off the issue when a provided fix has been confirmed to solve the issue (usually by the reporter) and has been merged.
I will have a quick look at PR @88 as well as the other outstanding PRs and merge asap.

**Comment by pfg666** [comment]
Sure. I will then wait for the PR to be merged before closing the issue.

**Comment by boaks** [comment]
Fixed with PR #109 

@pfg666 

Just in the case you're interested in verifying the solution, please do so.
Otherwise, please close the issue.

**Comment by pfg666** [comment]
Yeap, I can confirm that this fixed it. Closing the issue!
